### DISCUSSION TITLE
Proposal: list-patterns on enumerables

### DISCUSSION THREAD

**Discussion by jcouv** [dotnet/csharplang#9004] [Language Ideas]
Allow `is [ 1, 2, 3 ]` (list pattern) and `is [ 1, ..var x, 5 ]` (slice pattern) on enumerables.

Champion issue: https://github.com/dotnet/csharplang/issues/9005

Background:
- https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/list-patterns.md
- https://github.com/dotnet/csharplang/blob/main/proposals/csharp-8.0/ranges.md

Design meetings:
- https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-10-19.md
- https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-10-09.md#list-patterns-on-enumerables

**Comment by Xyncgas** [comment]
why not ```[1; x]``` ```[1; x; 5]```
since it's a new language proposal, we can throw away that ```var```, similar to how we can write lambda without type name on its parameter, it just matches
maybe we can even do ```var myLambda  = [1; x] => true | unmatched => false; myLambda(myList);```

**Comment by KennethHoff** [comment]
It's not a new language feature. It already exists for indexable collections, like arrays and lists.

**Comment by En3Tho** [comment]
[1; x] feels like a (personally) long awaited pattern matching on non const values rather that x being a declared variable here. Also, as stated above it's already a feature on indexable types. It's just that it's harder to do properly with sequences because it most likely will result in implicit buffering in complex cases like [1, ..var x, 5] which might be not desirable. At least with indexable collections this behavior is explicit and based on public members of those collections.

**Comment by jnm2** [comment]
Proposal revamped: <https://github.com/dotnet/csharplang/pull/9165/files?short_path=842dc5b#diff-842dc5be20dd67a91231b72c300d0bc365040809d32c98dee0e4e166e53191bc>

**Comment by smarts** [comment]
I know that languages have to walk the fine line between "pit of success" for developers and providing useful/powerful features that have potential for misuse. I could go either way on this one, but my question to the community is which side of the line people fall on for this feature?
Is it useful? Of course.
Does it encourage anti-patterns? At the very least it's making the multi-enumeration scenario easier, which is usually frowned upon because of deferred execution. At its worst it's relying on repeatable enumeration which isn't guaranteed for purely deferred enumerables nor collection-backed enumerables (because non-indexed collections should be seen as unordered and therefore don't need to provide repeatable enumeration - Note: a new interface might help here, but that's a separate issue). 
Is it more useful than harmful? Probably, but I just wanted to _enumerate_ some of the concerns. (pun intended)

**Reply by jnm2** [reply]
The proposal does guarantee that multiple enumeration is not introduced within the pattern match itself. As to whether the user chooses to introduce multiple enumeration by pattern-matching and then enumerating again, is it really different from the user writing `if (enumerable.Any())` and then enumerating again? The .NET SDK and ReSharper already have analysis rules which warn about this, and they would equally be able to see and catch it for pattern matching on the enumerable.

**Reply by smarts** [reply]
> As to whether the user chooses to introduce multiple enumeration by pattern-matching and then enumerating again, is it really different from the user writing if (enumerable.Any()) and then enumerating again?

No, but the `.Any()` or `.FirstOrDefault()` cases are pretty innocuous for the most part in my opinion. It's patterns like the examples in the description that are more worrrisome and making multiple enumeration easier.

**Reply by jnm2** [reply]
The `.Any()` example actually still comes with all the performance downsides and correctness issues, because it may be redoing a remote query and coming back with different results each time. What makes `is [p1, .., p2]` worse than `.Any()`?

**Reply by smarts** [reply]
Setting aside the "remote query" part for a second… Yes, the slice pattern can be much worse than `.Any()` if you consider a very large amount of items in the collection. I.e., O(n) operations vs. O(1). And just to reiterate, I'm not saying this can't already happen, only that it makes it easier/more convenient to do. I liken it to seeing `.ToArray()` and `.ToList()` peppered throughout code when it could have been changed in one place upstream and avoided everywhere else, or in some cases it's simply not needed at all. It's convenient _and_ very readable (which is a good thing), but that's also a double-edge sword.

**Reply by HaloFour** [reply]
>  Yes, the slice pattern can be much worse than `.Any()` if you consider a very large amount of items in the collection.

That depends on the pattern.  `Any()` is equivalent to `[_, ..]`, and only needs to ensure that a single element exists, so the size of the result isn't relevant.  Similarly, a pattern like `["foo", .., var last]` can bail immediately if the enumerable doesn't have a first element that is a `string` with the value of `"foo"`.

**Reply by timcassell** [reply]
I think the only issue would be if it's used in a switch expression. The compiler would need to ensure all switch arms being checked will only enumerate once. As opposed to `if else if`, which I assume would necessitate multiple enumeration.

**Reply by jnm2** [reply]
> Yes, the slice pattern can be much worse than `.Any()` if you consider a very large amount of items in the collection.

@smarts The proposal recommends against allowing subpatterns on the slice pattern for this reason. `..` in the pattern will not materialize a new collection for the items it matches against.

> I think the only issue would be if it's used in a switch expression. The compiler would need to ensure all switch arms being checked will only enumerate once. As opposed to `if else if`, which I assume would necessitate multiple enumeration.

@timcassell This is an excellent point, I'd like to understand whether there are guarantees like this for switches currently. I'm not sure there are. If there are, that would mean that the refactoring from `switch` to `if` could introduce multiple enumeration where it wasn't before.

**Reply by jnm2** [reply]
I'm not sure whether the following scenarios could realistically be implemented without multiple enumeration. I'm leaning towards letting it be multiple enumeration since there are multiple patterns, and just letting analyzers point this out:
```cs
obj switch
{
    { Enumerable: [], OtherProp: 1 } => ...
    { Enumerable: [1] or [2, 3], OtherProp: 4 } => ...
}
```
Let's just let this be enumeration three times, and you don't do this unless you'd do the same with `.Any()` and `.SequenceEqual` etc calls.

**Reply by jnm2** [reply]
It sounds like it might be possible to guarantee no multiple enumeration across switch branches and `or`/`and` patterns. That does make it more interesting.

**Reply by HaloFour** [reply]
I guess that depends on how complicated the compiler wanted to get.  It's already capable of avoiding reevaluating patterns that appear anywhere within the `switch` more than one time, I don't see why it couldn't technically handle that combination of patterns with a single enumeration that attempts to advance at most 3 times.  It could even decide that evaluating the `OtherProp` patterns would make it easier/cheaper to apply the list patterns.
